test/api_test.py

```diff
@@ -148,3 +148,75 @@ def test_delete_devices(self):
 
         with pytest.raises(MatrixRequestError):
             self.cli.api.delete_devices(self.auth_body, [self.device_id])
+
+
+class TestKeysApi:
+    cli = client.MatrixClient("http://example.com")
+    user_id = "@alice:matrix.org"
+    device_id = "JLAFKJWSCS"
+    one_time_keys = {"curve25519:AAAAAQ": "/qyvZvwjiTxGdGU0RCguDCLeR+nmsb3FfNG3/Ve4vU8"}
+    device_keys = {
+        "user_id": "@alice:example.com",
+        "device_id": "JLAFKJWSCS",
+        "algorithms": [
+            "m.olm.curve25519-aes-sha256",
+            "m.megolm.v1.aes-sha"
+        ],
+        "keys": {
+            "curve25519:JLAFKJWSCS": "3C5BFWi2Y8MaVvjM8M22DBmh24PmgR0nPvJOIArzgyI",
+            "ed25519:JLAFKJWSCS": "lEuiRJBit0IG6nUf5pUzWTUEsRVVe/HJkoKuEww9ULI"
+        },
+        "signatures": {
+            "@alice:example.com": {
+                "ed25519:JLAFKJWSCS": ("dSO80A01XiigH3uBiDVx/EjzaoycHcjq9lfQX0uWsqxl2gi"
+                                       "MIiSPR8a4d291W1ihKJL/a+myXS367WT6NAIcBA")
+            }
+        }
+    }
+
+    @responses.activate
+    @pytest.mark.parametrize("args", [
+        {},
```

Comment:
I don't really entirely understand pytest. Where is this empty dict going? `upload_keys` only takes two arguments, not three.

Reply:
https://docs.pytest.org/en/latest/parametrize.html  
The test is called 3 times, one with {} which stands for no argument, then with device_keys set, then with one_time_keys. Not sure why I didn't do a fourth test with both set. Then we call the API method with the arguments found in `args`.  
If this is too confusing for what it is, I can go back to copy-pasting, which won't make much difference seeing the rest of the file :)